exporter/collector/metricsexporter.go

```diff
@@ -298,3 +369,197 @@ func mergeLabels(mergeInto labels, others ...labels) labels {
 
 	return mergeInto
 }
+
+// Takes a GCM metric type, like (workload.googleapis.com/MyCoolMetric) and returns the display name.
+func (m *metricMapper) metricTypeToDisplayName(mURL string) string {
+	// TODO - user configuration around display name?
+	// Default: strip domain, keep path after domain.
+	u, err := url.Parse(fmt.Sprintf("metrics://%s", mURL))
+	if err != nil {
+		return mURL
+	}
+	return strings.TrimLeft(u.Path, "/")
+}
+
+// Returns label descriptors for a metric.
+func (m *metricMapper) labelDescriptors(pm pdata.Metric) []*label.LabelDescriptor {
+	// TODO - allow customization of label descriptions.
+	result := []*label.LabelDescriptor{}
+	switch pm.DataType() {
+	case pdata.MetricDataTypeGauge:
+		points := pm.Gauge().DataPoints()
+		for i := 0; i < points.Len(); i++ {
+			points.At(i).Attributes().Range(func(key string, _ pdata.AttributeValue) bool {
+				result = append(result, &label.LabelDescriptor{
+					Key: key,
+				})
+				return true
+			})
+		}
+	case pdata.MetricDataTypeSum:
+		points := pm.Sum().DataPoints()
+		for i := 0; i < points.Len(); i++ {
+			points.At(i).Attributes().Range(func(key string, _ pdata.AttributeValue) bool {
+				result = append(result, &label.LabelDescriptor{
+					Key: key,
+				})
+				return true
+			})
```

Comment:
nit could add a closure func which takes `pdata.AttributeMap` and adds the label descriptors to `result` to reduce some of this duplication. Don't think it's possible to get rid of the switch/for loops though :(

Reply:
This was, quite honestly, the most painful thing to write in Go so far, knowing that abstraction COULD have occurred.

That cleaned up the worst of it, still wish there was an abstract way to pull attributes from subpoints without knowing specific structures.